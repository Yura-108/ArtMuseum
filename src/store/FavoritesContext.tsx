import { useFavorites } from '@utils/hooks/useFavorites.ts';
import React, { createContext, useContext } from 'react';
import { FavoriteType } from '@types/favoriteTypes.ts';

const FavoritesContext = createContext<FavoriteType | null>(null);

export const FavoritesProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const favorites = useFavorites();
  return (
    <FavoritesContext.Provider value={favorites}>
      {children}
    </FavoritesContext.Provider>
  );
};

export const useFavoritesContext = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error(
      'useFavoritesContext must be used within a FavoritesProvider',
    );
  }
  return context;
};
