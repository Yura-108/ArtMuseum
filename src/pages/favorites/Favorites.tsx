import React from 'react';
import favoriteIcon from '@assets/images/favoriteIcon.svg';
import ContainerSmallCards from '@components/ContainerSmallCards/ContainerSmallCard.tsx';
import ErrorMessage from '@components/ErrorMessage/ErrorMessage.tsx';
import Message from '@components/Message/Message.tsx';
import SectionTitle from '@components/SectionTitle/SectionTitle.tsx';
import SmallCardContainerSkeleton from '@components/Skeletons/SmallCardContainerSkeleton.tsx';
import Title from '@components/Title/Title.tsx';
import { useFavoritesContext } from '@store/FavoritesContext.tsx';
import { useQuery } from '@tanstack/react-query';
import { getArtWorks } from '@utils/API/APIFunctions.ts';
import ErrorBoundary from '@components/ErrorBoundary/ErrorBoundary.tsx';

const Favorites: React.FC = () => {
  const { favorites } = useFavoritesContext();
  const { data, isPending, isError, error } = useQuery({
    queryKey: ['page', favorites],
    queryFn: () => getArtWorks(favorites),
  });

  if (isError) return <ErrorMessage>{error.message}</ErrorMessage>;

  return (
    <section>
      <Title>
        here are your <br />
        <div>
          <img src={favoriteIcon} alt="favorite icon" /> <span>Favorites</span>
        </div>
      </Title>
      <SectionTitle h2={'Your favorites list'} h4={'Saved by you'} />
      {isPending ? (
        <SmallCardContainerSkeleton length={favorites.length} />
      ) : data.length ? (
        <ErrorBoundary fallback={<h2>Mistake: failed to load components</h2>}>
          <ContainerSmallCards data={data} />
        </ErrorBoundary>
      ) : (
        <Message>You don't have any favorite artworks yet</Message>
      )}
    </section>
  );
};

export default Favorites;
