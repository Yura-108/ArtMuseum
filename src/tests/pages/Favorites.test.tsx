import { render, screen } from '@testing-library/react';
import { useFavoritesContext } from '@store/FavoritesContext';
import { useQuery } from '@tanstack/react-query';
import Favorites from '@pages/favorites/Favorites.tsx';

jest.mock('@store/FavoritesContext.tsx', () => ({
  useFavoritesContext: jest.fn(),
}));

jest.mock('@tanstack/react-query', () => ({
  useQuery: jest.fn(),
}));

jest.mock('@components/ContainerSmallCards/ContainerSmallCard.tsx', () => ({ data }: any) => (
  <div>Rendered {data.length} favorite cards</div>
));
jest.mock('@components/ErrorMessage/ErrorMessage.tsx', () => ({ children }: any) => <div>{children}</div>);
jest.mock('@components/Message/Message.tsx', () => ({ children }: any) => <div>{children}</div>);
jest.mock('@components/Skeletons/SmallCardContainerSkeleton.tsx', () => ({ length }: any) => (
  <div>Loading {length} cards...</div>
));
jest.mock('@components/ErrorBoundary/ErrorBoundary.tsx', () => ({ fallback, children }: any) => (
  <div>{children || fallback}</div>
));

describe('Favorites', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders loading skeleton when data is loading', () => {
    (useFavoritesContext as jest.Mock).mockReturnValue({ favorites: [1, 2, 3] });
    (useQuery as jest.Mock).mockReturnValue({ data: null, isPending: true, isError: false });

    render(<Favorites />);
    expect(screen.getByText('Loading 3 cards...')).toBeInTheDocument();
  });

  it('renders error message when an error occurs', () => {
    (useFavoritesContext as jest.Mock).mockReturnValue({ favorites: [1, 2, 3] });
    (useQuery as jest.Mock).mockReturnValue({
      data: null,
      isPending: false,
      isError: true,
      error: { message: 'Failed to fetch data' },
    });

    render(<Favorites />);
    expect(screen.getByText('Failed to fetch data')).toBeInTheDocument();
  });

  it('renders "no favorites" message when the list is empty', () => {
    (useFavoritesContext as jest.Mock).mockReturnValue({ favorites: [] });
    (useQuery as jest.Mock).mockReturnValue({ data: [], isPending: false, isError: false });

    render(<Favorites />);
    expect(screen.getByText("You don't have any favorite artworks yet")).toBeInTheDocument();
  });

  it('renders favorite cards when data is available', () => {
    const mockData = [{ id: 1, title: 'Art 1' }, { id: 2, title: 'Art 2' }];
    (useFavoritesContext as jest.Mock).mockReturnValue({ favorites: [1, 2] });
    (useQuery as jest.Mock).mockReturnValue({ data: mockData, isPending: false, isError: false });

    render(<Favorites />);
    expect(screen.getByText('Rendered 2 favorite cards')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const mockData = [{ id: 1, title: 'Art 1' }];
    (useFavoritesContext as jest.Mock).mockReturnValue({ favorites: [1] });
    (useQuery as jest.Mock).mockReturnValue({ data: mockData, isPending: false, isError: false });

    const { asFragment } = render(<Favorites />);
    expect(asFragment()).toMatchSnapshot();
  });
});
