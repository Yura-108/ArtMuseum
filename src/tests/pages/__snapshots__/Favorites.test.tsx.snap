// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites matches snapshot 1`] = `
<DocumentFragment>
  <section>
    <h1>
      here are your 
      <br />
      <div>
        <img
          alt="favorite icon"
        />
         
        <span>
          Favorites
        </span>
      </div>
    </h1>
    <div
      class="containerTitle"
    >
      <h4>
        Saved by you
      </h4>
      <h2>
        Your favorites list
      </h2>
    </div>
    <div>
      <div>
        Rendered 1 favorite cards
      </div>
    </div>
  </section>
</DocumentFragment>
`;
